flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -131,8 +131,13 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
 
 	@Override
 	protected void cancelTask() {
-		if (headOperator != null) {
-			headOperator.cancel();
+		try {
+			if (headOperator != null) {
+				headOperator.cancel();
+			}
+		}
+		finally {
+			sourceThread.interrupt();
```

Comment:
I'd restore that `sourceThread` interrupted status (just for clarity, if someone later will modify it's catch statement).
Currently, it catches `Throwable`, completes future with it and exits.

Reply:
Could elaborate/rephrase what do you mean?

Reply:
I mean to add `catch InterruptedException` section before `catch Throwable` in `LegacySourceFunctionThread` (with `Thread.currentThread().interrupt();` in it).